### PR TITLE
Feature: Implement auto renewal load tester 

### LINKED ISSUE #1: Auto-renewal load tester
In a programming language of your choice (I recommend Go) write a program that takes:

-host HOST -port PORT -cacert CA_CERT_FILE -cakey CA_KEY_FILE -concurrency AMOUNT -sleep SECONDS -duration DAYS

It shall:

1. Create a private key (as Icinga 2 would do) in memory
2. Create AMOUNT certificates for the private key valid for DAYS days signed by CA_*_FILE (as Icinga 2 would do) in memory; endpoint names: 1, 2, … AMOUNT
3. At the same time connect AMOUNT times to HOST:PORT via TLS (don’t verify the peer) and use another of the newly issued certificates every time
4. Once connected, sleep SECONDS seconds and request a new certificate (as Icinga 2 would do, you already know where to cherry-pick JSON-RPC stuff from)

### PR DESCRIPTION
**Pull request by yhabteab**
resolves #1

### CODE REVIEW

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Don't generate anything temporary where something concrete is required.

File: main.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+			Certificates: []tls.Certificate{cert},
+			ClientAuth: tls.RequireAnyClientCert,
+			ClientCAs: caCertPool,
+			InsecureSkipVerify: true,
+		}
+
+		conf.Rand = rand.Reader
+		Addr := fmt.Sprintf("%s%s%d", *host, ":", *port)
+
+		conn, err := tls.Dial("tcp", Addr, conf)
```
Comment (Al2Klimov) [inline main.go]: The task was: "At the same time connect AMOUNT times", i.e. in parallel.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	block, _ := pem.Decode(rootCAs)
+	if block == nil {
+		log.Fatalf("CLIENT: Failed to parse certificate PEM")
+	}
+	CAuthority, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to parse certificate: %s", err.Error())
+	}
+
+	for i := 0; i < *concurrency; i++ {
```
Comment (Al2Klimov) [inline main.go:82]: But you're connecting one after the other.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		conf := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth: tls.RequireAnyClientCert,
+			ClientCAs: caCertPool,
+			InsecureSkipVerify: true,
+		}
+
+		conf.Rand = rand.Reader
+		go ProcessOutgoingConnection(host, port, conf)
+		time.Sleep(*sleep)
```
Comment (Al2Klimov) [inline main.go]: Now you're even sleeping between the connection attempts.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+		conf := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth: tls.RequireAnyClientCert,
+			ClientCAs: caCertPool,
+			InsecureSkipVerify: true,
+		}
+
+		conf.Rand = rand.Reader
+		go ProcessOutgoingConnection(host, port, conf)
```
Comment (Al2Klimov) [inline main.go]: That makes this parallelism here pretty useless.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
The task was to first generate the certificates (and I'd do the other preparation incl. `conf.Rand = rand.Reader` there as well) **and then** connect multiple times in parallel.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		}
+
+		conf := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth: tls.RequireAnyClientCert,
+			ClientCAs: caCertPool,
+			InsecureSkipVerify: true,
+		}
+
+		conf.Rand = rand.Reader
```
Comment (Al2Klimov) [inline main.go]: Didn't I suggest to put this preparation in the first loop?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+	log.Printf("CLIENT: JsonRpcConnection: Sending pki::RequestCertificate JSRPC message: %s", message)
+
+	err = base.WriteNetStringToStream(conn, message)
+
+	if err != nil {
+		log.Printf("CLIENT: JsonRpcConnection: Error while sending json encoded JSRPC message: %s", err)
+	}
+
+	time.Sleep(*sleep)
```
Comment (Al2Klimov) [inline main.go]: Please lookup in the task description **between which two** operations to sleep.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		NotBefore:      time.Now(),
+		NotAfter:       time.Now().AddDate(0, 0, *days),
+		SubjectKeyId: []byte{1, 2, 3, 4, 6},
+		KeyUsage:       x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:    []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		IsCA:           false,
+		IPAddresses:    []net.IP{net.ParseIP("127.0.0.1"), net.IPv6loopback},
+	}
+
+	certBytes, err := x509.CreateCertificate(rand.Reader, &certificateTemplate, CACert, &priv.PublicKey, CAKey)
```
Comment (Al2Klimov) [inline main.go]: Please dump a such certificate (probably in an extra program) and verify that this #1 task is done properly:

"Create (...) certificates (...) (**as Icinga 2 would do**)"

**Comment by yhabteab** [comment]
Except for the ``notAfter`` and ``notBefore`` time offsets, it looks exactly like the certificates created by Icinga2.

**Icinga2**:
``openssl x509 -in ~/Workspace/icinga2/prefix/var/lib/icinga2/certs/ha-master.crt -text``

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            16:04:b3:ac:39:47:09:91:a3:df:01:ac:d6:d5:c7:a5:f4:56:3a:29
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = Icinga CA
        Validity
            Not Before: Dec 15 17:00:00 2020 GMT
            Not After : Dec 12 17:00:00 2035 GMT
        Subject: CN = ha-master
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (4096 bit)
                Modulus:
                    00:a9:7d:21:19:57:77:15:2d:61:9d:11:b1:73:03:
                    [...]
                    cb:45:a6:31:cb:9a:c3:30:e8:fc:08:4f:24:4d:e5:
                    a2:3f:7f
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name: 
                DNS:ha-master
    Signature Algorithm: sha256WithRSAEncryption
[...]
```

**Test Server**:
``openssl x509 -in ha-cluster.crt -text``

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            d7:cc:48:b3:ee:d5:7b:7b:23:71:fc:c5:ce:9c:00:c5:24:32:64:3f
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = Icinga CA
        Validity
            Not Before: Jan 11 10:11:39 2021 GMT
            Not After : Feb 10 10:11:39 2021 GMT
        Subject: CN = localhost2
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (4096 bit)
                Modulus:
                    00:bd:f1:4e:52:b8:97:20:c9:61:66:8f:bb:99:e1:
                    [...]
                    20:4b:15:b2:0c:e3:2a:a8:6d:99:80:8d:3d:03:ec:
                    57:8c:33
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name: 
                DNS:localhost2
    Signature Algorithm: sha256WithRSAEncryption
[...]
```

File: main.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	block, _ := pem.Decode(rootCAs)
+	if block == nil {
+		log.Fatalf("CLIENT: Failed to parse certificate PEM")
+	}
+	CAuthority, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to parse certificate: %s", err.Error())
+	}
+
+	tlsConfigs := make(map[int]*tls.Config)
```
Comment (Al2Klimov) [inline main.go:80]: Why a map?

Comment (yhabteab) [reply main.go:80]: Is easier for me than slice array.

File: main.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+	for i := 0; i < *concurrency; i++ {
+		go ProcessOutgoingConnection(host, port, tlsConfigs[i], sleep)
+	}
+
+	fmt.Scanln()
+}
+
+func ProcessOutgoingConnection(host *string, port *int, conf *tls.Config, sleep *time.Duration)  {
+	Addr := fmt.Sprintf("%s%s%d", *host, ":", *port)
```
Comment (Al2Klimov) [inline main.go]: Consider doing this, ...

File: main.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+	encoder := &Encoder{
+		JSONRPC: "2.0",
+		Method: "pki::RequestCertificate",
+		Params: map[string]string{
+			"ticket": " ",
+		},
+	}
+
+	message, err := json.Marshal(&encoder)
```
Comment (Al2Klimov) [inline main.go]: ... and this only once for all.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 10 above)
Comment (yhabteab) [reply main.go]: I do not understand what exactly you mean!

Comment (Al2Klimov) [reply main.go]: Addr and message are equal for all (let’s say) 100 connections, right?

Comment (yhabteab) [reply main.go]: yeah exactly.

Comment (Al2Klimov) [reply main.go]: And why do you compute the very same values 100 times?

Comment (yhabteab) [reply main.go]: Do you want me to encode the message in the main function and pass it as a parameter to the ``ProcessOutgoingConnection()``  function?

Comment (Al2Klimov) [reply main.go]: ... and the same for Addr.

File: main.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+	defer conn.Close()
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to connect: %s", err.Error())
+	}
+
+	log.Printf("CLIENT: Connect to %s succeed", Addr)
+	time.Sleep(*sleep)
+
+	message, err := json.Marshal(&encoder)
```
Comment (Al2Klimov) [inline main.go]: You’re still encoding the message inside the function.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+type ConfigInfo map[string]string
+
+func main()  {
+	port := flag.Int("port", 0, "On which port the server starting to listen")
+	host := flag.String("host", " ", "Host Name or IP Address")
+	cakey := flag.String("cakey", " ", "Private Key Certificate Chain")
+	cacert := flag.String("cacert", " ", "Certificate Authority")
+	concurrency := flag.Int("concurrency", 0, "concurrency")
+	sleep := flag.Duration("sleep", 0, "sleep in seconds")
+	duration := flag.Int("duration", 0, "Amount of files to be synced to the specified zone")
```
Comment (Al2Klimov) [inline main.go]: Re-read the task, especially what VALUES the program shall take via which -params and what the VALUES are for.

Are you sure about the descriptions, especially of -port and -duration?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Comment by Al2Klimov** [comment]
Test whether #2 fulfills #1 by putting this program instead of the agent in a master<->agent cluster. Share the master's logs.

**Comment by yhabteab** [comment]
```powershell
./auto-loader -port 5665 -host "127.0.0.1" -cakey /Users/yhabteab/Workspace/icinga2/prefix/var/lib/icinga2/ca/ca.key -cacert /Users/yhabteab/Workspace/icinga2/prefix/var/lib/icinga2/ca/ca.crt -concurrency 3 -sleep 10s -duration 30
```

```powershell
[2021-01-11 16:13:09 +0100] information/ApiListener: New client connection for identity 'localhost0' from [::ffff:127.0.0.1]:61399 (no Endpoint object found for identity)
[2021-01-11 16:13:09 +0100] information/ApiListener: New client connection for identity 'localhost1' from [::ffff:127.0.0.1]:61398 (no Endpoint object found for identity)
[2021-01-11 16:13:09 +0100] information/ApiListener: New client connection for identity 'localhost2' from [::ffff:127.0.0.1]:61400 (no Endpoint object found for identity)
[2021-01-11 16:13:13 +0100] information/WorkQueue: #5 (ApiListener, RelayQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-01-11 16:13:13 +0100] information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2021-01-11 16:13:20 +0100] information/JsonRpcConnection: Received certificate request for CN 'localhost0' signed by our CA.
[2021-01-11 16:13:20 +0100] information/JsonRpcConnection: Received certificate request for CN 'localhost1' signed by our CA.
[2021-01-11 16:13:20 +0100] information/JsonRpcConnection: Received certificate request for CN 'localhost2' signed by our CA.
[2021-01-11 16:13:20 +0100] information/JsonRpcConnection: Sending certificate response for CN 'localhost1' to endpoint 'localhost1'.
[2021-01-11 16:13:20 +0100] information/JsonRpcConnection: Sending certificate response for CN 'localhost2' to endpoint 'localhost2'.
[2021-01-11 16:13:20 +0100] information/JsonRpcConnection: Sending certificate response for CN 'localhost0' to endpoint 'localhost0'.
```

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**